File: server/datasets/java_source-code/Multiple-G2/structural_35_avl_tree.java
class Node_10 {
    int k_10, h_10;
    Node_10 left, right;
    Node_10(int d) {
        k_10 = d;
        h_10 = 1;
    }
}
public class Tree_10 {
    public static void main(String[] args) {
        Tree_10 t = new Tree_10();
        Node_10 r = null;
        r = t.ins_10(r, 10);
    }
    int max(int a, int b) {
        return (a > b) ? a : b;
    }
    int h_10(Node_10 N) {
        if (N == null) return 0;
        return N.h_10;
    }
    int gb_10(Node_10 N) {
        if (N == null) return 0;
        return h_10(N.left) - h_10(N.right);
    }
    Node_10 upd_10(Node_10 n_10) {
        n_10.h_10 = 1 + max(h_10(n_10.left), h_10(n_10.right));
        return n_10;
    }
    Node_10 rr_10(Node_10 y) {
        Node_10 x = y.left;
        Node_10 T2 = x.right;
        x.right = y;
        y.left = T2;
        upd_10(y);
        upd_10(x);
        return x;
    }
    Node_10 rl_10(Node_10 x) {
        Node_10 y = x.right;
        Node_10 T2 = y.left;
        y.left = x;
        x.right = T2;
        upd_10(x);
        upd_10(y);
        return y;
    }
    Node_10 ins_10(Node_10 n_10, int k_10) {
        if (n_10 == null) return (new Node_10(k_10));
        if (k_10 < n_10.k_10) n_10.left = ins_10(n_10.left, k_10);
        else if (k_10 > n_10.k_10) n_10.right = ins_10(n_10.right, k_10);
        else return n_10;
        upd_10(n_10);
        int b_10 = gb_10(n_10);
        if (b_10 > 1 && k_10 < n_10.left.k_10) return rr_10(n_10);
        if (b_10 < -1 && k_10 > n_10.right.k_10) return rl_10(n_10);
        if (b_10 > 1 && k_10 > n_10.left.k_10) {
            n_10.left = rl_10(n_10.left);
            return rr_10(n_10);
        }
        if (b_10 < -1 && k_10 < n_10.right.k_10) {
            n_10.right = rr_10(n_10.right);
            return rl_10(n_10);
        }
        return n_10;
    }
}

File: server/datasets/python_source-code/Multiple-G2/structural_35_avl_tree.py
class Node_10:
    def __init__(self, v_10):
        self.v_10 = v_10
        self.r_10 = None
        self.l_10 = None
        self.height = 1

def h_10(node):
    if not node:
        return 0
    return node.height

def b_10(node):
    if not node:
        return 0
    return h_10(node.l_10) - h_10(node.r_10)

def rr_10(y):
    x = y.l_10
    sub = x.r_10
    x.r_10 = y
    y.l_10 = sub
    y.height = 1 + max(h_10(y.l_10), h_10(y.r_10))
    x.height = 1 + max(h_10(x.l_10), h_10(x.r_10))
    return x

def rl_10(x):
    y = x.r_10
    sub = y.l_10
    y.l_10 = x
    x.r_10 = sub
    x.height = 1 + max(h_10(x.l_10), h_10(x.r_10))
    y.height = 1 + max(h_10(y.l_10), h_10(y.r_10))
    return y

def ins_10(rt_10, v_10):
    if not rt_10:
        return Node_10(v_10)
    if v_10 < rt_10.v_10:
        rt_10.l_10 = ins_10(rt_10.l_10, v_10)
    elif v_10 > rt_10.v_10:
        rt_10.r_10 = ins_10(rt_10.r_10, v_10)
    else:
        return rt_10

    rt_10.height = 1 + max(h_10(rt_10.l_10), h_10(rt_10.r_10))
    bf_10 = b_10(rt_10)

    if bf_10 > 1 and v_10 < rt_10.l_10.v_10:
        return rr_10(rt_10)
    if bf_10 < -1 and v_10 > rt_10.r_10.v_10:
        return rl_10(rt_10)
    if bf_10 > 1 and v_10 > rt_10.l_10.v_10:
        rt_10.l_10 = rl_10(rt_10.l_10)
        return rr_10(rt_10)
    if bf_10 < -1 and v_10 < rt_10.r_10.v_10:
        rt_10.r_10 = rr_10(rt_10.r_10)
        return rl_10(rt_10)

    return rt_10
